packages/react-native/ReactCommon/react/performance/timeline/PerformanceEntryReporter.h

```diff
@@ -144,54 +100,57 @@ class PerformanceEntryReporter {
       const std::optional<std::string>& startMark = std::nullopt,
       const std::optional<std::string>& endMark = std::nullopt);
 
+  // https://w3c.github.io/user-timing/#clearmarks-method
+  // https://w3c.github.io/user-timing/#clearmeasures-method
   void clearEntries(
       std::optional<PerformanceEntryType> entryType = std::nullopt,
       std::string_view entryName = {});
 
-  std::vector<PerformanceEntry> getEntries(
-      std::optional<PerformanceEntryType> entryType = std::nullopt,
-      std::string_view entryName = {}) const;
-
-  void logEventEntry(
-      std::string name,
-      double startTime,
-      double duration,
-      double processingStart,
-      double processingEnd,
-      uint32_t interactionId);
-
-  void logLongTaskEntry(double startTime, double duration);
-
-  const std::unordered_map<std::string, uint32_t>& getEventCounts() const {
-    return eventCounts_;
-  }
-
-  DOMHighResTimeStamp getCurrentTimeStamp() const;
-
-  void setTimeStampProvider(std::function<double()> provider) {
-    timeStampProvider_ = std::move(provider);
-  }
-
  private:
-  std::function<void()> callback_;
+  std::unique_ptr<PerformanceObserverRegistry> observerRegistry_;
 
-  mutable std::mutex entriesMutex_;
-  std::array<PerformanceEntryBuffer, NUM_PERFORMANCE_ENTRY_TYPES> buffers_;
-  std::unordered_map<std::string, uint32_t> eventCounts_;
+  mutable std::mutex buffersMutex_;
+  PerformanceEntryCircularBuffer eventBuffer_{EVENT_BUFFER_SIZE};
+  PerformanceEntryCircularBuffer longTaskBuffer_{LONG_TASK_BUFFER_SIZE};
+  PerformanceEntryKeyedBuffer markBuffer_;
+  PerformanceEntryKeyedBuffer measureBuffer_;
 
-  uint32_t droppedEntriesCount_{0};
+  std::unordered_map<std::string, uint32_t> eventCounts_;
 
   std::function<double()> timeStampProvider_ = nullptr;
 
-  mutable std::mutex nameLookupMutex_;
-
   double getMarkTime(const std::string& markName) const;
-  void scheduleFlushBuffer();
 
-  void getEntries(
-      PerformanceEntryType entryType,
-      std::string_view entryName,
-      std::vector<PerformanceEntry>& res) const;
+  inline PerformanceEntryBuffer& getBufferRef(PerformanceEntryType entryType) {
+    switch (entryType) {
+      case PerformanceEntryType::EVENT:
+        return eventBuffer_;
+      case PerformanceEntryType::MARK:
+        return markBuffer_;
+      case PerformanceEntryType::MEASURE:
+        return measureBuffer_;
+      case PerformanceEntryType::LONGTASK:
+        return longTaskBuffer_;
+      default:
+        assert(0 && "Unhandled PerformanceEntryType");
+    }
+  }
+
+  const inline PerformanceEntryBuffer& getBuffer(
```

Comment:
Do we really need 2 versions of this method?

Reply:
Yep, for const correctness. This one is used most of the times from methods marked as `const`, and the second one (ref) is used in `clear`, which is mutating